### DISCUSSION TITLE
Target errors in pipeline, runs in errored workspace

### DISCUSSION THREAD

**Discussion by pvtodorov** [ropensci/targets#991] [Help] [answered]
I keep getting an error within a pipeline, but upon diving down into the errored workspace, the target executes with no issues. While the larger pipeline is too big to share, I have an example with a single target from it. All targets of this type error in the same way and I am looking for hints into what to do next.

An `.RData` image of the workspace is available [nhgc2_MOL___all.obob5v5__v__all.obobBL6.RData](https://1drv.ms/u/s!AosQ4TBamCdUhqRnSVxBtbiClL_rEg?e=CH3O0n)

The target being run is `nhgc2_MOL___all.obob5v5__v__all.obobBL6`

the output of manifest limited to targets containing this name
```r
tar_manifest() %>% filter(str_detect(name, 'nhgc2_MOL___all.obob5v5__v__all.obobBL6')) %>% print
```
```
  name                                       command                                                                         
  <chr>                                      <chr>                                                                           
1 nhgc2_MOL___all.obob5v5__v__all.obobBL6    nhg2cell(nhm_MOL___obob5v5, da_results_nhg_MOL___all.obob5v5__v__all.obobBL6)   
2 nhgc2_MOL___all.obob5v5__v__all.obobBL6d5  nhg2cell(nhm_MOL___obob5v5, da_results_nhg_MOL___all.obob5v5__v__all.obobBL6d5) 
3 nhgc2_MOL___all.obob5v5__v__all.obobBL6d14 nhg2cell(nhm_MOL___obob5v5, da_results_nhg_MOL___all.obob5v5__v__all.obobBL6d14)
```

the output from meta limited to errored targets containing this name (I did not wait for the other two to run and error in this case)
```r
tar_meta() %>% filter(!is.na(error)) %>% filter(str_detect(name, 'nhgc2_MOL___all.obob5v5__v__all.obobBL6')) %>% select(name, error) %>% print
```
```
  name                                    error                
  <chr>                                   <chr>                
1 nhgc2_MOL___all.obob5v5__v__all.obobBL6 object name not found
```

running the pipeline
```r
tar_make(names=c("nhgc2_MOL___all.obob5v5__v__all.obobBL6"), shortcut=TRUE)
```
```
• start target nhgc2_MOL___all.obob5v5__v__all.obobBL6
• record workspace nhgc2_MOL___all.obob5v5__v__all.obobBL6
✖ error target nhgc2_MOL___all.obob5v5__v__all.obobBL6
✖ error target nhgc2_MOL___all.obob5v5__v__all.obobBL6
• end pipeline [19.846 seconds]
```
opening the errored target workspace and executing the command manually with success
```r
> tar_workspaces()
[1] "nhgc2_MOL___all.obob5v5__v__all.obobBL6"
> tar_workspace(nhgc2_MOL___all.obob5v5__v__all.obobBL6)
> nhg2cell(nhm_MOL___obob5v5, da_results_nhg_MOL___all.obob5v5__v__all.obobBL6)
# A tibble: 3,968 × 3
   rowname              restored_grouping fgf1_grouping
   <chr>                <fct>             <chr>
 1 CTCCTCCTCAAGTAAG_3_2 pos_away          pos
 2 TTCCTTCTCGACGATT_1_3 pos_away          pos
 3 CGAGGCTAGTGTACCT_4_3 pos_away          pos
 4 CTCAATTCAAGGGCAT_4_1 pos_FGF1          pos
 5 CATCCCATCCGTGTGG_4_1 pos_FGF1          pos
 6 TGGAGGATCTCTTAAC_4_2 pos_FGF1          pos
 7 TCATGAGGTGTGTCGC_4_2 pos_FGF1          pos
 8 TGAGGAGCAAACCATC_5_2 pos_FGF1          pos
 9 GAGACTTTCATTCTTG_5_2 pos_FGF1          pos
10 GAGGCCTAGATGTAGT_5_2 pos_FGF1          pos
```

subset of the network visualized, showing upstream dependencies are up to date and exist
```r
tar_visnetwork(names=c("nhgc2_MOL___all.obob5v5__v__all.obobBL6",
                       "nhm_MOL___obob5v5",
                       "da_results_nhg_MOL___all.obob5v5__v__all.obobBL6",
                       "nhg2cell"),
               shortcut=TRUE)
```
![image](https://user-images.githubusercontent.com/13072605/204559495-bf778c49-6bf6-419b-af91-e158b6ae5474.png)

**Comment by wlandau** [comment, marked as answer]
In that case, I suggest debugging in other ways, e.g. the [`debug` option](https://books.ropensci.org/targets/debugging.html#pause-the-pipeline-with-the-debug-option) or `browser()` in a [function](https://books.ropensci.org/targets/debugging.html#pause-the-pipeline-with-browser). And it is also a good idea to [peel back layers](https://books.ropensci.org/targets/debugging.html#system-issues), e.g. see if the error still happens with `tar_make()` (as opposed to `tar_make_clustermq()` or `tar_make_future()`). 

Workspace files are lightweight references to other parts of your [data store](https://books.ropensci.org/targets/data.html#local-data-store), so unfortunately I cannot troubleshoot. If you can, please post inline code I can run to reproduce the error. As a general rule, I do not open embedded data files because of the potential security concerns.

**Reply by pvtodorov** [reply]
Fair! I will give these options a shot

**Comment by pvtodorov** [comment]
I looked at the function being called, this is one of the few places where "name" from the error occurs in my code. In this instance, the intent is to use `rename()` from `dplyr`

Changing a line from
```r
rename(Nhood = name) %>%
``` 
to 
```r
dplyr::rename(Nhood = name) %>%
```
actually resolves the issue.

I am not sure why this is happening but the pipeline has been fragile with respect to dplyr and I've had to specify that in several other places explicitly. 

In the target script I load the following libraries
```r
library(targets)
library(tarchetypes) # Load other packages as needed. # nolint
library(future)
library(future.callr)
```
I do not define any packages under `tar_option_set()` and for the most part the tidyverse makes it across into the targets as I don't need to specify given libraries explicitly.

What exactly is going on? Why does it work in some cases but not others?

**Reply by pvtodorov** [reply]
just as interstingly – when I open the errored workspace `rename` is the correct one
```r
> rename
```
```
function (.data, ...)
{
    UseMethod("rename")
}
<bytecode: 0x55e73e5bf1a0>
<environment: namespace:dplyr>
>
```

**Reply by wlandau** [reply]
`dplyr::rename()` may conflict with `rename()` from other packages, so the order that packages get loaded matters. 

Before running its command, each target loads its packages in the order listed in the `packages` argument of `tar_target()`. The default value of the `packages` argument `tar_option_get("packages")`, which is set with `tar_option_set(packages = c("..."))`. The default of `tar_option_get("packages")` is `(.packages())`: whatever happens to be loaded in `_targets.R` at the time the target is defined.

In `tar_make()`, each target gets run in the same R session that runs `_targets.R`. But in `tar_make_clustermq()` and `tar_make_future()`, different targets run in different R sessions, so the `packages` argument does matter in those cases.

Namespacing with `::` is a good workaround too.

**Reply by pvtodorov** [reply]
Thanks for your patience yet again!

I have dug deeper and it's become apparent that one of the scripts sourced into my `_targets.R` loads `library(SummarizedExperiment)` which then loads `S4Vectors` that has its own `rename` method.

I believe this may be stomping the `dplyr::rename()` method from `tidyverse`. 

Still not 100% clear on why this is the case. All imports in `_targets.R`, including sourcing of other scripts happens before any targets are declared. Still mysterious why it would have the `dplyr::rename()` when loading the workspace.

I get these errors running the pipeline with `tar_make()` and not `tar_make_future()`. Does having the future and callr  loaded in `_targets.R` cause `tar_make()` to run differently?
```r
# Load packages required to define the pipeline:
library(targets)
library(tarchetypes) # Load other packages as needed. # nolint
library(future)
library(future.callr)
plan(callr)
```

My `tar_options_set()` is also set to do IO on the workers. Would this affect how `tar_make()` behaves when running without parallelization?
```r
# Set target options:
tar_option_set(
#   packages = c("tibble"), # packages that your targets need to run
  format = "qs", # default storage format,
  error = "null",
  retrieval = "worker",
  storage = "worker"
  # Set other options as needed.
)
```

Nevertheless I will try to replicate a minimal example now that I know what fixes the error.

**Reply by wlandau** [reply]
> Still not 100% clear on why this is the case. All imports in _targets.R, including sourcing of other scripts happens before any targets are declared. Still mysterious why it would have the dplyr::rename() when loading the workspace.

It has to do with the order that the packages are loaded. One way selects `dplyr::rename()`, the other selects `SummarizedExperiment::rename()`.

> I get these errors running the pipeline with tar_make() and not tar_make_future(). 

`tar_make_future()` creates a new R process for each target, whereas `tar_make()` runs all the targets in the same R session. In the former case, the package order is determined by the `packages` argument. In the latter, many of the the packages in the `packages` argument are already loaded by the time the target runs, and reloading them will not change the order of precedence. In other words, if you set packages with `library()` instead of `tar_option_set()`, then `tar_make()` and `tar_make_future()` may load packages in a different order. That's why it works in one case and not the other.

> My tar_options_set() is also set to do IO on the workers. Would this affect how tar_make() behaves when running without parallelization?

`storage` and `retrieval` are only relevant in `tar_make_clustermq()` and `tar_make_future()`.
